### PR TITLE
[Defender (RiskIQ) - EASM] Api version 2023-03-01-preview

### LINKED ISSUE #22683: [Defender (RiskIQ) - EASM] API Review
New API Review meeting has been requested.

**Service Name**: RiskIQ - EASM
**Review Created By**: Brian Zak
**Review Date**: 3/21/2023 1:00 PM PT
**PR**: 
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: This meeting is to review the addition of a 'Data Connector' API that is moving from private preview to public preview.

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=9673a9bf-37ae-ed11-aad1-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by thang-bit**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document: https://msazure.visualstudio.com/RiskIQ/_git/easm-dataconnections?path=/docs/Architecture.md&version=GBmain&_a=preview&anchor=data-plane-api
* Previous Open API Doc: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/riskiq/data-plane/Microsoft.Easm/preview/2022-11-01-preview/easm.json
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

**Comment by JeffreyRichter** [comment]
This swagger looks like an ARP/RP swagger. If it is, you need to contact the ARM review board.

**Comment by thang-bit** [comment]
Although the pathing makes it look "arm-ish", this swagger is data-plane

**Comment by thang-bit** [comment]
@DominikMe / @JeffreyRichter can we have a review on this PR?

**Comment by JeffreyRichter** [comment]
Isn't this an ARM/control plane API? If so, we don't review these, the ARM team does.

**Comment by thang-bit** [comment]
@JeffreyRichter this is our data plane API. The fact that it has subscriptions, resource groups, etc, makes it look like ARM. The reason that our data plane looks like ARM is because when our API interact with MySQL database, we need to get some data from ARM. Therefore, it was recommended to us to follow the similar pattern of control plane

**Comment by JeffreyRichter** [comment]
OK, I'll review it. Can you point me to anything specifically to look at as the full API is pretty big.

**Comment by thang-bit** [comment]
Great! @JeffreyRichter  This is the summary of things you should review:
- Data Connections paths
- Paging response models
- Disco groups validate endpoint

Just incase that you want to have more context of the changes and initial feedbacks, this is the link for the record of the last API Review meeting: [[RiskIQ - EASM] API Review - Issue _22683-20230321_160403-Meeting Recording.mp4](https://microsoft.sharepoint.com/teams/AzureAPIStewardshipBoard/_layouts/15/stream.aspx?id=%2Fteams%2FAzureAPIStewardshipBoard%2FShared%20Documents%2FGeneral%2FRecordings%2F%5BRiskIQ%20%2D%20EASM%5D%20API%20Review%20%2D%20Issue%20%5F22683%2D20230321%5F160403%2DMeeting%20Recording%2Emp4&ct=1683651420971&or=Teams-HL&ga=1&LOF=1)

**Comment by JeffreyRichter** [comment]
I started to watch the video again to refresh myself on the issues and I was not in this meeting at all. So, I think you should get other people to review this as they'd have more context than I do.

**Comment by thang-bit** [comment]
@JeffreyRichter I can't change the reviewers on this PR. Who should I contact to change the reviewers?

**Comment by JeffreyRichter** [comment]
How did I get added in the first place?

Contact Mark Weitzel

-- Jeffrey
________________________________

**Comment by thang-bit** [comment]
@JeffreyRichter You are automatically assigned, but I can't change the reviewers. I will contact Mark then
image

**Comment by mikekistler** [comment]
I have switched reviewers to myself and Ted Glaza.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (mikekistler) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:1546]: This property and many others in this API definition has no description. All properties must have a clear and meaningful description.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (mikekistler) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:419]: Do not use the HTTP method in the operationId. A PUT that only returns 200 should use the verb "replace".  If it is possible for this operation to also _Create_ a DataConnnection the a 201 response should be defined and the operationId should use "CreateOrReplace".

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (mikekistler) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:658]: As above, don't use "Put" in the operationId.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (mikekistler) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:1361]: As above, don't use "Put" in the operationId.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (mikekistler) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:1739]: The `value` property of a pageable response should be _required_:
```suggestion
      },
      "required": ["value"],
```

**Review by mikekistler — COMMENTED** [review]
I left a few quick comments based on the Spectral linter. These are mostly minor and do not need to be fixed for preview but _must_ be fixed for GA.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (thang-bit) [reply specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:419]: @mikekistler Should we still use "Get" and "Delete" in the operationId? Like "DataConnections_Get" and "DataConnections_Delete"

**Comment by thang-bit** [comment]
@mikekistler among these suggestions, which are mandatory for public preview?

**Comment by thang-bit** [comment]
@mikekistler I have noted the recommendations above to address them before the GA review

**Review by mikekistler — APPROVED** [review]
Looks good for preview. 👍

**Comment by thang-bit** [comment]
@mikekistler, Can we change the other reviewer, I think Ted Glaza is out of office

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (weidongxu-microsoft) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:2523]: the "total" or "count" is not recommended.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (weidongxu-microsoft) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:642]: For a PUT, wondering whether same roundtrip model can be shared by both request and response (if property only in response, mark as readonly).

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (weidongxu-microsoft) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:403]: same question, can it be roundtrip model?

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (weidongxu-microsoft) [inline specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:2415]: nit, the name appear not consistent. One use suffix Date, another At.

**Comment by thang-bit** [comment]
@weidongxu-microsoft Do these issues also not need to be fixed for previews, but must be fixed for GA?

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (thang-bit) [reply specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:642]: Can you please elaborate on or give me a documentation for the roundtrip model?

Comment (weidongxu-microsoft) [reply specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:642]: https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#rest-response-body-is-resource-schema

example (just click the first one I found) https://github.com/Azure/azure-rest-api-specs/blob/main/specification/purview/data-plane/Azure.Analytics.Purview.Account/preview/2019-11-01-preview/account.json#L225-L243

**Comment by weidongxu-microsoft** [comment]
> @weidongxu-microsoft Do these issues also not need to be fixed for previews, but must be fixed for GA?

No, they probably don't block your preview.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (thang-bit) [reply specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:642]: I don't think we'd use roundtrip model here since there are a few properties only available in response, but we'll definitely mark them as readonly

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (thang-bit) [reply specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:403]: Same answer as DiscoGroup

**Comment by thang-bit** [comment]
@weidongxu-microsoft I have noted all the comments, can we have an approval on this PR please?

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (weidongxu-microsoft) [reply specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:642]: "readonly" do mean customer cannot set the value, which is the standard solution to the case of "a few properties only available in response".

Every single ARM resource is such, that `id` property is readonly, and cannot set by user.

**Comment by weidongxu-microsoft** [comment]
I've mentioned my comment won't block you, but I will leave the sign off to Mike.

File: specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json
Comment (mikekistler) [reply specification/riskiq/data-plane/Microsoft.Easm/preview/2023-03-01-preview/easm.json:642]: You should only mark properties as readOnly if they are in a schema that could be sent in a request body. If the schema is only used for responses, there is no value in marking properties readOnly.

**Review by mikekistler — APPROVED** [review]
I discovered a few more issues. The one on the LRO pattern is particularly concerning -- our SDK generators will not give a very good result here. But for preview we can probably live with it.  Please add these to the list to fix for your next preview (you should do another preview given the number of changes needed).

**Pull request merged**
